File: src/main/java/ru/pasha__kun/windowfx/FxmlWindow.java
package ru.pasha__kun.windowfx;

import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.fxml.FXMLLoader;
import ru.pasha__kun.windowfx.loader.WindowLoader;

import java.io.IOException;
import java.net.URL;

public class FxmlWindow extends Window {
    private final ObjectProperty<FXMLLoader> loader = new SimpleObjectProperty<>(this, "loader");
    private final ObjectProperty<URL> base = new SimpleObjectProperty<>(this, "base");

    @Override
    public void onLoad(WindowContainer container) throws WindowLoadException {
        super.onLoad(container);
        WindowLoader l = container.getWindowLoader();
        if (l == null) throw new IllegalStateException("FxmlWindow must be added with loader!");

        try {
            FXMLLoader loader = this.loader.get();
            loader.setLocation(new URL(base.get(), l.toResourceName(getClass().getSimpleName()) + ".fxml"));
            setContent(loader.load());
        } catch (IOException e){
            throw new WindowLoadException(e);
        }
    }

    public FXMLLoader getLoader() {
        return loader.get();
    }

    public ObjectProperty<FXMLLoader> loaderProperty() {
        return loader;
    }

    public void setLoader(FXMLLoader loader) {
        this.loader.set(loader);
    }
}

File: src/main/kotlin/FxmlWindow.kt
package ru.pasha__kun.windowfx

import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.fxml.FXMLLoader
import java.io.IOException
import java.net.URL

//open class FxmlWindow : Window() {
//    val loaderProperty: ObjectProperty<FXMLLoader> = SimpleObjectProperty(this, "loader")
//    val baseProperty: ObjectProperty<URL> = SimpleObjectProperty(this, "base")
//
//    var loader: FXMLLoader by loaderProperty
//    var base: URL by baseProperty
//
//    @Throws(WindowLoadException::class)
//    override fun onLoad(container: WindowsContainer?) {
//        super.onLoad(container)
//        val l = container!!.windowLoader ?: throw IllegalStateException("FxmlWindow must be added with loader!")
//        try {
//            val loader = loader
//            loader.location = URL(base, l.toResourceName(javaClass.simpleName) + ".fxml")
//            content = loader.load()
//        } catch (e: IOException) {
//            throw WindowLoadException(e)
//        }
//    }
//}